cpp/ql/src/semmle/code/cpp/dataflow/internal/SubBasicBlocks.qll

```diff
@@ -98,16 +102,22 @@ class SubBasicBlock extends ControlFlowNodeBase {
    * start from 0, and the node at position 0 always exists and compares equal
    * to `this`.
    */
+  pragma[nomagic]
   ControlFlowNode getNode(int pos) {
-    exists(BasicBlock bb | bb = this.getBasicBlock() |
-      exists(int thisPos | this = bb.getNode(thisPos) |
-        result = bb.getNode(thisPos + pos) and
-        pos >= 0 and
-        pos < this.getNumberOfNodes()
+    exists(BasicBlock bb |
+      exists(int outerPos |
+        result = bb.getNode(outerPos) and
+        pos = outerPosToInnerPos(bb, outerPos)
       )
     )
   }
 
+  pragma[nomagic]
+  private int outerPosToInnerPos(BasicBlock bb, int posInBB) {
+    posInBB = result + this.getIndexInBasicBlock(bb) and
+    result = [ 0 .. this.getNumberOfNodes() - 1 ]
```

Comment:
How confident are we that this isn't quadratic?  Naively, it seems that for any particular node `bb, posInBB`, every possible `this` in `bb` must be searched.

Other than this question, FWIW, these changes LGTM also.

Reply:
This is the DIL we generate for this predicate:

```
noinline nomagic
SubBasicBlocks::SubBasicBlock::outerPosToInnerPos#ffff(/* SubBasicBlocks::SubBasicBlock */ cached entity this,
                                                       /* BasicBlocks::BasicBlock */ cached entity bb,
                                                       int posInBB, int result) :-
  exists(cached int call_result |
    exists(int range#high |
      exists(int Sub#lhs |
        SubBasicBlocks::SubBasicBlock::getIndexInBasicBlock#fff(this, bb,
                                                                call_result),
        SubBasicBlocks::SubBasicBlock::getNumberOfNodes#ff(this, Sub#lhs),
        Minus(Sub#lhs, 1, range#high)
      ),
      range(0, range#high, result)
    ),
    PlusInt(result, call_result, posInBB)
  )
.
```

It first joins in the two getters (functional relations), so the tuple count will be the number of `SubBasicBlock`. It then joins with the range, which multiplies that tuple count by the average number of nodes per `SubBasicBlock`. All together, this predicate should have the same number of tuples as there are `ControlFlowNode`s.

Reply:
Great, that gives me a bit more confidence.